### PR TITLE
Offline course sync people list and details

### PR DESCRIPTION
**Pull request by idevainstr**
refs: MBL-16766
affects: Student
release note: None.
test plan: 
- Sync People
- Check listed people
- Check user details for current and other users
- Search, filter, and message send button should be disabled in offline

## Checklist

- [ ] Follow-up e2e test ticket created
- [ ] A11y checked
- [ ] Tested on phone
- [ ] Tested on tablet
- [ ] Tested in dark mode
- [ ] Tested in light mode
- [ ] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>One or more files are below the minimum test coverage 50%</td>
    </tr>
  </tbody>
</table>


#### Release Note: 
None.

#### Affected Apps: Student

[MBL-16766](https://instructure.atlassian.net/browse/MBL-16766)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 91.9% | 92.41% | -0.51%
Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift | 0% | -- | --
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 43fd201874380675336fa47b144b56c149c9b131
</p>

**Comment by balintbartok** [comment]
please add unit tests for CourseSyncPeopleInteractor.swift

**Comment by balintbartok** [comment]
IMG_0018
I tried everything (reset cache, fresh install), but couldn't see people details. 
Also, after syncing and closing the app, the next time it insta-crashes :(

**Review by vargaat — CHANGES_REQUESTED** [review]
There's no option to sync People in the Offline Content selector screen.

File: Core/Core/People/ContextUser.swift [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        let user: ContextUser = context.first(where: #keyPath(ContextUser.id), equals: item.id.value) ?? context.insert()
+        user.id = item.id.value
+        user.name = item.name
+        user.shortName = item.short_name
+        user.sortableName = item.sortable_name
+        user.email = item.email
+        user.avatarURL = item.avatar_url?.rawValue
+        user.pronouns = item.pronouns
+        if let enrollments = item.enrollments {
+            for item in enrollments {
```
Comment (vargaat) [inline Core/Core/People/ContextUser.swift:51]: It's a little confusing that the function's property is called `item` and the local variable within this loop is also called `item`.

**Review by vargaat — CHANGES_REQUESTED** [review]
Functionality wise looks good. I had a few code quality remarks.

File: Core/Core/People/ContextUser.swift [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    @NSManaged public var id: String
+    @NSManaged public var name: String
+    @NSManaged public var sortableName: String
+    @NSManaged public var shortName: String
+    @NSManaged public var avatarURL: URL?
+    @NSManaged public var email: String?
+    @NSManaged public var courseID: String?
+    @NSManaged public var groupID: String?
+    @NSManaged public var pronouns: String?
+    @NSManaged public var observerID: String?
```
[the remark this answers is not in the document]
Comment (idevainstr) [reply Core/Core/People/ContextUser.swift:32]: removed.

File: Core/Core/People/ContextUser.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+public final class ContextUser: NSManagedObject {
+    @NSManaged public var id: String
+    @NSManaged public var name: String
+    @NSManaged public var sortableName: String
+    @NSManaged public var shortName: String
+    @NSManaged public var avatarURL: URL?
+    @NSManaged public var email: String?
+    @NSManaged public var courseID: String?
+    @NSManaged public var groupID: String?
```
[the remark this answers is not in the document]
Comment (idevainstr) [reply Core/Core/People/ContextUser.swift:30]: removed.

File: Core/Core/People/ContextUser.swift [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import CoreData
+
+public final class ContextUser: NSManagedObject {
+    @NSManaged public var id: String
+    @NSManaged public var name: String
+    @NSManaged public var sortableName: String
+    @NSManaged public var shortName: String
+    @NSManaged public var avatarURL: URL?
+    @NSManaged public var email: String?
+    @NSManaged public var courseID: String?
```
[the remark this answers is not in the document]
Comment (idevainstr) [reply Core/Core/People/ContextUser.swift:29]: removed.

File: Core/Core/People/ContextUser.swift (hunk 1 above)
Comment (idevainstr) [reply Core/Core/People/ContextUser.swift:51]: renamed.

File: Core/Core/People/ContextCard/Course/ContextCardViewModel.swift [hunk 5]
```diff
@@ -41,28 +41,31 @@ public class ContextCardViewModel: ObservableObject {
...
         self.userID = userID
         self.context = Context.course(courseID)
         self.isViewingAnotherUser = (userID != currentUserID)
         self.isSubmissionRowsVisible = isSubmissionRowsVisible
         self.isLastActivityVisible = isLastActivityVisible
         self.isModal = isModal
+        self.offlineModeViewModel = offlineModeViewModel
+
+        offlineModeViewModel.objectWillChange.sink { [weak self] _ in self?.objectWillChange.send() }
+            .store(in: &subscriptions)
```
[the remark this answers is not in the document]
Comment (idevainstr) [reply Core/Core/People/ContextCard/Course/ContextCardViewModel.swift:62]: done.

File: Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    private static func fetchCourse(context: Context) -> AnyPublisher<Void, Error> {
+        ReactiveStore(useCase: GetCourse(courseID: context.id))
+            .getEntities()
+            .mapToVoid()
+            .eraseToAnyPublisher()
+    }
+
+    private static func fetchSubmissionsForStudent(context: Context, userID: String) -> AnyPublisher<Void, Error> {
+        guard userID == AppEnvironment.shared.currentSession?.userID else { return Just(()).setFailureType(to: Error.self).eraseToAnyPublisher() }
```
Comment (szabinst) [inline Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift:78]: I think a new line after/before `else` would make it more readable.

File: Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+                    $0.enrollment_state == .active &&
+                    $0.user_id.value == userID
+                }
+                if let apiEnrollment = apiEnrollment, let id = apiEnrollment.id?.value {
+                    let databaseContext = env.database.viewContext
+                    let enrollment: Enrollment = databaseContext.first(where: #keyPath(Enrollment.id), equals: id) ?? databaseContext.insert()
+                    enrollment.update(fromApiModel: apiEnrollment, course: nil, in: databaseContext)
+                }
+                promise(.success(()))
+            }}
```
Comment (szabinst) [inline Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift:144]: Couldn't we create a use case for this logic?

File: Core/CoreTests/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractorLiveTests.swift [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+@testable import Core
+import XCTest
+
+class CourseSyncPeopleInteractorLiveTests: CoreTestCase {
+    private let testee = CourseSyncPeopleInteractorLive()
```
Comment (szabinst) [inline Core/CoreTests/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractorLiveTests.swift:23]: I'd make this a var and recreate it during `setUp` and nil it out in the `tearDown` function to be on the safe side.

**Review by szabinst — CHANGES_REQUESTED** [review]
It doesn't work for me when I select multiple courses. It looks like it only downloads for one course. 

When I select one course only, it works as expected. 

https://github.com/instructure/canvas-ios/assets/110813321/5963d0a9-8ad7-4f8c-94be-2a08d2f19a84

File: Core/Core/People/ContextCard/Course/ContextCardViewModel.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (idevainstr) [reply Core/Core/People/ContextCard/Course/ContextCardViewModel.swift:62]: for real this time.

File: Core/CoreTests/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractorLiveTests.swift (hunk 8 above)
Comment (idevainstr) [reply Core/CoreTests/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractorLiveTests.swift:23]: done.

File: Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift (hunk 6 above)
Comment (idevainstr) [reply Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift:78]: done

File: Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift (hunk 7 above)
Comment (idevainstr) [reply Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncPeopleInteractor.swift]: and we did.

**Pull request closed without merging**